File: 151-reverse-words-in-a-string/151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        Stack<String>st=new Stack<>();
        StringBuilder sb=new StringBuilder();
        
        for(String ss:s.trim().split(" ")){
            if(!ss.isEmpty())st.push(ss);
        }
        
        while(st.size()>0){
            sb.append(st.pop());
            sb.append(" ");
        }
        
        return sb.toString().trim();
        
    }
}

File: 151-reverse-words-in-a-string/151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        reverse(s.begin(),s.end());
        
        int l=0;
        int r=0;
        int n=s.size();
        int i=0;
        
        while(i<n){
            
            while(i<n && s[i]!=' '){
                s[r++]=s[i++];
            }
            
            if(l<r){
                reverse(s.begin()+l , s.begin()+r);
                
                if(r==n)break;
                
                s[r++]=' ';
                l=r;
            }
            
            i++;
        }
        
        if(r>0 && s[r-1]==' ')r--;
        
        s.resize(r);
        
        return s;
    }
};